python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+        if position < 0 or position > self.size:
+            raise ValueError('position argument is out-of-range')
+        buf = source if isinstance(source, Buffer) else py_buffer(source)
+
+        if nbytes < 0:
+            # copy from host buffer to device buffer starting from
+            # position until device buffer is full
+            nbytes_ = min(self.size - position, buf.size)
+        else:
+            if nbytes > buf.size:
+                raise ValueError(
+                    'requested more to copy than available from host buffer')
+            if nbytes > self.size - position:
+                raise ValueError(
+                    'requested more to copy than available in device buffer')
+            # copy nbytes from host buffer to device buffer starting
+            # from position
+            nbytes_ = nbytes
+
+        cdef shared_ptr[CBuffer] buf_ = pyarrow_unwrap_buffer(buf)
+        check_status(self.cuda_buffer.get().
+                     CopyFromHost(position, buf_.get().data(), nbytes_))
+        return nbytes_
+
+    def export_for_ipc(self):
+        """
+        Expose this device buffer as IPC memory which can be used in other
+        processes.
+
+        After calling this function, this device memory will not be
+        freed when the CudaBuffer is destructed.
+
+        Results
+        -------
+        ipc_handle : CudaIpcMemHandle
+          The exported IPC handle
+
+        """
+        cdef shared_ptr[CCudaIpcMemHandle] handle
+        check_status(self.cuda_buffer.get().ExportForIpc(& handle))
+        return pyarrow_wrap_cudaipcmemhandle(handle)
+
+    @property
+    def context(self):
+        """Returns the CUDA driver context of this buffer.
+        """
+        return pyarrow_wrap_cudacontext(self.cuda_buffer.get().context())
+
+    def slice(self, offset=0, length=None):
+        """Return slice of device buffer
+
+        Parameters
+        ----------
+        offset : int, default 0
+          Specify offset from the start of device buffer to slice
+        length : int, default None
+          Specify the length of slice (default is until end of device
+          buffer starting from offset)
+
+        Returns
+        -------
+        sliced : CudaBuffer
+          Zero-copy slice of device buffer.
+        """
+        if offset < 0 or offset >= self.size:
+            raise ValueError('offset argument is out-of-range')
+        cdef int64_t offset_ = offset
+        cdef int64_t size
+        if length is None:
+            size = self.size - offset_
+        elif offset + length <= self.size:
+            size = length
+        else:
+            raise ValueError(
+                'requested larger slice than available in device buffer')
+        parent = pyarrow_unwrap_cudabuffer(self)
+        return pyarrow_wrap_cudabuffer(
+            shared_ptr[CCudaBuffer](make_shared[CCudaBuffer](parent,
+                                                             offset_, size)))
```

Comment:
This suggests we should add a virtual `Slice` method to buffers which can be overridden. The default would return the current behavior of `arrow::SliceBuffer`

Reply:
Created ARROW-3221